### PR TITLE
fix: unit tests 

### PR DESCRIPTION
**Pull request by Ad96el**
Some clean-ups:

Fixed-point numbers are rounded. When minting, the requested amount and the calculated collateral are rounded up, while during burning, they are rounded down. This ensures that there is always sufficient collateral during burning.

Some errors have been mapped.

### CODE REVIEW

File: pallets/pallet-bonded-coins/src/lib.rs [hunk 1]
```diff
@@ -595,30 +602,33 @@ pub mod pallet {
...
 				target_currency_id.clone(),
-				&beneficiary,
+				&who,
 				amount_to_burn,
 				WithdrawalPrecision::Exact,
 				Fortitude::Force,
 			)?;
 
-			if !pool_details.transferable {
+			let user_funds = T::Fungibles::balance(target_currency_id.clone(), &who);
```
Comment (Ad96el) [inline pallets/pallet-bonded-coins/src/lib.rs:619]: Todo: Change this.

**Review by rflechtner — COMMENTED** [review]
The two rounding functions have very similar tasks, can we give them similar names?

File: pallets/pallet-bonded-coins/src/curves/mod.rs [hunk 2]
```diff
@@ -151,6 +155,13 @@ where
 	// This function can panic in theory, but only if frac_nbits() would be larger
 	// than 256 - and no Fixed of that size exists.
 	x_u256.shl_assign(CurveParameterTypeOf::<T>::frac_nbits());
+
+	if round_kind == &Round::Up {
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/curves/mod.rs]: added in: [abcbbb6](https://github.com/KILTprotocol/kilt-node/pull/808/commits/abcbbb6d7e1bc5796da94774da95234634c5e63c)

File: pallets/pallet-bonded-coins/src/curves/mod.rs [hunk 3]
```diff
@@ -164,3 +175,34 @@ where
 	// Return the result of scaled conversion to fixed.
 	Ok(fixed)
 }
+
+/// Rounds the given value to the given rounding kind.
+pub(crate) fn round<T: Config>(
+	value: CurveParameterTypeOf<T>,
+	round_kind: &Round,
+	denomination: u8,
+) -> Result<CollateralCurrenciesBalanceOf<T>, ArithmeticError>
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/curves/mod.rs]: I renamed the function in: [abcbbb6](https://github.com/KILTprotocol/kilt-node/pull/808/commits/abcbbb6d7e1bc5796da94774da95234634c5e63c)

File: pallets/pallet-bonded-coins/src/curves/mod.rs [hunk 4]
```diff
@@ -164,3 +175,34 @@ where
...
+
+	let trailing_zeros = value_u256.trailing_zeros();
+
+	let frac_bits: u32 = CurveParameterTypeOf::<T>::frac_nbits().into();
+
+	value_u256.shr_assign(frac_bits);
+
+	if round_kind == &Round::Up && trailing_zeros < frac_bits {
+		value_u256 = value_u256.checked_add(U256::from(1)).ok_or(ArithmeticError::Overflow)?;
+	}
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/curves/mod.rs]: Added explanation in: [abcbbb6](https://github.com/KILTprotocol/kilt-node/pull/808/commits/abcbbb6d7e1bc5796da94774da95234634c5e63c)

File: pallets/pallet-bonded-coins/src/lib.rs [hunk 5]
```diff
@@ -280,6 +281,10 @@ pub mod pallet {
 		InvalidInput,
 		Internal,
 		Slippage,
+		/// The amount to mint or burn is below the minimum allowed.
+		BelowMinimum,
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/lib.rs]: [eee5ccf](https://github.com/KILTprotocol/kilt-node/pull/808/commits/eee5ccf9980434aa2bf1e7dfe2ba84a9de707733)

File: pallets/pallet-bonded-coins/src/lib.rs [hunk 6]
```diff
@@ -624,19 +661,28 @@ pub mod pallet {
...
 				target_currency_id.clone(),
-				&beneficiary,
+				&who,
 				amount_to_burn,
 				WithdrawalPrecision::Exact,
 				Fortitude::Force,
 			)?;
 
-			if !pool_details.transferable {
+			let user_funds = T::Fungibles::total_balance(target_currency_id.clone(), &who);
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/lib.rs]: Added in: [abcbbb6](https://github.com/KILTprotocol/kilt-node/pull/808/commits/abcbbb6d7e1bc5796da94774da95234634c5e63c)

File: pallets/pallet-bonded-coins/src/lib.rs [hunk 7]
```diff
@@ -497,11 +505,17 @@ pub mod pallet {
...
+			ensure!(pool_details.state.is_live(), Error::<T>::PoolNotLive);
 			ensure!(pool_details.can_mint(&who), Error::<T>::NoPermission);
 
 			let number_of_currencies = Self::get_currencies_number(&pool_details);
 			ensure!(number_of_currencies <= currency_count, Error::<T>::CurrencyCount);
 
+			ensure!(
+				amount_to_mint >= pool_details.min_operation_balance.saturated_into(),
+				Error::<T>::BelowMinimum
+			);
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/lib.rs]: Changed the order: [5b51a0f](https://github.com/KILTprotocol/kilt-node/pull/808/commits/5b51a0f7606d15225a076d04d2a3103161ca9b8c)

File: pallets/pallet-bonded-coins/src/lib.rs [hunk 8]
```diff
@@ -510,14 +524,20 @@ pub mod pallet {
 				.get(currency_idx)
 				.ok_or(Error::<T>::IndexOutOfBounds)?;
 
+			let round_kind = Round::Up;
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/lib.rs:629]: Yeah… I had it the same way, but then I thought putting the variant into a variable would be more resilient against developer errors.

File: pallets/pallet-bonded-coins/src/lib.rs [hunk 9]
```diff
@@ -394,6 +401,7 @@ pub mod pallet {
 			let pool_details = Pools::<T>::get(&pool_id).ok_or(Error::<T>::PoolUnknown)?;
 
 			ensure!(pool_details.is_manager(&who), Error::<T>::NoPermission);
+			ensure!(pool_details.state.is_live(), Error::<T>::PoolNotLive);
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/lib.rs:431]: Updated the order in: [5b51a0f](https://github.com/KILTprotocol/kilt-node/pull/808/commits/5b51a0f7606d15225a076d04d2a3103161ca9b8c)

File: pallets/pallet-bonded-coins/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (rflechtner) [reply pallets/pallet-bonded-coins/src/lib.rs:629]: we do make sure it is consistent, that is true... ok let's leave it

**Review by rflechtner — APPROVED** [review]
u 🪨 !

**Comment by ntn-x2** [comment]
Is it me or this PR also contains changes from a different PR? I see a new line added for the `CollateralIsZero` error, which should not be part of this test PR, right?

**Comment by Ad96el** [comment]
Yes, I combined two smaller PRs here. The changes are intendant.

**Pull request merged**